src/components/ThemeProvider/ThemeProvider.tsx

```diff
@@ -19,28 +21,73 @@ export function ThemeProvider({
   children,
 }: ThemeProviderProps) {
   const {unstableGlobalTheming = false} = useFeatures();
+
+  const parentContext = useContext(ThemeContext);
+  const isParentThemeProvider = parentContext === undefined;
+  const parentMode = parentContext && parentContext.mode && parentContext.mode;
+  const parentColors =
+    parentContext &&
+    parentContext.UNSTABLE_colors &&
+    parentContext.UNSTABLE_colors;
+
+  const {UNSTABLE_colors, mode, ...rest} = themeConfig;
+
+  const childShouldInheritParentColors =
+    !isParentThemeProvider && mode !== undefined && mode !== parentMode;
+
+  const defaultColors = {
+    surface: UNSTABLE_Color.Surface,
+    onSurface: UNSTABLE_Color.OnSurface,
+    interactive: UNSTABLE_Color.Interactive,
+    neutral: UNSTABLE_Color.Neutral,
+    primary: UNSTABLE_Color.Primary,
+    critical: UNSTABLE_Color.Critical,
+    warning: UNSTABLE_Color.Warning,
+    highlight: UNSTABLE_Color.Highlight,
+    success: UNSTABLE_Color.Success,
+    decorative: UNSTABLE_Color.Decorative,
+  };
+
+  const processedThemeConfig: ThemeConfig = {
+    ...rest,
+    ...{mode: mode || parentMode},
+    UNSTABLE_colors: {
+      ...(isParentThemeProvider && defaultColors),
+      ...(childShouldInheritParentColors && parentColors),
+      ...UNSTABLE_colors,
+    },
+  };
+
   const customProperties = useMemo(
-    () => buildCustomProperties(themeConfig, unstableGlobalTheming),
-    [unstableGlobalTheming, themeConfig],
+    () =>
+      buildCustomProperties(
+        processedThemeConfig,
+        unstableGlobalTheming,
+        isParentThemeProvider ? Tokens : undefined,
+      ),
+    [isParentThemeProvider, processedThemeConfig, unstableGlobalTheming],
   );
+
   const theme = useMemo(
     () =>
       buildThemeContext(
-        themeConfig,
+        processedThemeConfig,
         unstableGlobalTheming ? customProperties : undefined,
       ),
-    [customProperties, themeConfig, unstableGlobalTheming],
+    [customProperties, processedThemeConfig, unstableGlobalTheming],
   );
 
-  // We want these values to be `null` instead of `undefined` when not set.
+  // We want these values to be empty string instead of `undefined` when not set.
```

Comment:
Probably fine to leave this change, `null` seems to behave the same way but maybe this is safer for browser support? Alternatively, this could be more explicit and call `document.body.style.removeProperty` when the custom property is `undefined`.

Reply:
In TS 3.7 the updated the types of the builtins so null isn't considered valid and it threw an compile exception. I updated this in the update to TS3.7 but it got reverted. https://github.com/Shopify/polaris-react/pull/2237/files#diff-ce476db90622695c5696acdd862caf42

Reply:
Thanks for the refresher!